UpdatePublishedVersions.ps1

```diff
@@ -0,0 +1,149 @@
+    [Parameter(Mandatory=$true)][string]$nupkgPath,
+    # Print out the new file contents, but don't change the versions repository.
+    [switch]$dryRun)
+
+function ConvertPathTo-Package([string]$path)
+{
+    # Find the package ID and version using a regex. This matches the semantic version
+    # and assumes everything to the left is the id or a path to the package directory.
+    $matched = $path -match '^(.*\\)?(.*?)\.(([0-9]+\.)?[0-9]+\.[0-9]+(-([A-z0-9-]+))?)\.(symbols\.)?nupkg$'
+    if ($matched)
+    {
+        $packageInfo = @{
+            Path = $path
+            Name = $matches[2]
+            Version = $matches[3]
+            Prerelease = $matches[6]
+        }
+        $packageInfo.NameVersion = "$($packageInfo.Name) $($packageInfo.Version)"
+        return $packageInfo
+    }
+    else
+    {
+        throw "Couldn't find name and version from path $path."
+    }
+}
+
+# Updates a GitHub file with the specified file contents
+function Update-GitHub-File(
+    [string]$user = $gitHubUser,
+    [string]$email = $gitHubEmail,
+    [string]$authToken = $gitHubAuthToken,
+    [string]$owner = $versionsRepoOwner,
+    [string]$repo = $versionsRepo,
+    [string]$path,
+    [string]$newFileContent,
+    [string]$commitMessage)
+{
+    function message([string]$message)
+    {
+        Write-Host -ForegroundColor Green "*** $message ***"
+    }
+
+    $headers = @{
+        'Accept' = 'application/vnd.github.v3+json'
+        'Authorization' = "token $authToken"
+    }
+
+    $fileUrl = "https://api.github.com/repos/$owner/$repo/contents/$path"
+
+    message "Getting the `"sha`" of the current contents of file '$owner/$repo/$path'"
+
+    $currentFile = Invoke-WebRequest $fileUrl -UseBasicParsing -Headers $headers
+    $currentSha = (ConvertFrom-Json $currentFile.Content).sha
+
+    message "Got `"sha`" value of '$currentSha'"
+
+    message "Request to update file '$owner/$repo/$path' contents to:"
+    Write-Host $newFileContent
+
+    if ($dryRun)
+    {
+        message 'Not sending request: dry run.'
+        return
+    }
+
+    # Base64 encode the string
+    function ToBase64([string]$value)
+    {
+       return [System.Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($value))
+    }
+
+    $updateFileBody = 
+@"
+{
+  "message": "$commitMessage",
+  "committer": {
+    "name": "$user",
+    "email": "$email"
+  },
+  "content": "$(ToBase64 $newFileContent)",
+  "sha": "$currentSha"
+}
+"@
+
+    message 'Sending request...'
+    $putResponse = Invoke-WebRequest $fileUrl -UseBasicParsing -Method PUT -Body $updateFileBody -Headers $headers
+
+    if ($putResponse.StatusCode -ge 200 -and $putResponse.StatusCode -lt 300)
+    {
+        message 'Successfully updated the file'
+    }
+}
+
+try
+{
+    $packages = Get-ChildItem $nupkgPath | %{ ConvertPathTo-Package $_ }
+}
+catch
+{
+    # Any exception in parsing terminates.
+    throw $_
```

Comment:
This behavior is actually even more bizarre than I thought. This code continues past exceptions:
```
$packages = Get-ChildItem $nupkgPath | %{ ConvertPathTo-Package $_ }
```
But this code stops at the exception (in `ConvertPathTo-Package`):
```
$packagePaths = Get-ChildItem $nupkgPath
$packages = $packagePaths | %{ ConvertPathTo-Package $_ }
```
It looks like `Get-ChildItem` is the source of the weirdness, because piping a list also stops on the exception:
```
$packages = @('Microsoft.CSharp.4..nupkg') | %{ ConvertPathTo-Package $_ }
```
I'll change this to use an intermediate variable instead of the try-catch, but I wasn't able to find a way to fix this as a single assignment.